core/src/kura.rs

```diff
@@ -620,6 +728,67 @@ impl BlockStore {
         Ok(())
     }
 
+    /// Write the hash of a single block at the specified `block_height`.
+    /// If `block_height` is beyond the end of the index file, attempt to
+    /// extend the index file.
+    ///
+    /// # Errors
+    /// IO Error.
+    pub fn write_block_hash(
+        &mut self,
+        block_height: u64,
+        hash: HashOf<VersionedCommittedBlock>,
+    ) -> Result<()> {
+        let path = self.path_to_blockchain.join(HASHES_FILE_NAME);
+        let mut hashes_file = std::fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(path.clone())
+            .map_err(|e| Error::IO(e, path.clone()))?;
+        let start_location = block_height * SIZE_OF_BLOCK_HASH;
+        if start_location + SIZE_OF_BLOCK_HASH
+            > hashes_file
+                .metadata()
+                .map_err(|e| Error::IO(e, path.clone()))?
```

Comment:
I'm not sure my idea makes sense in this case, but I can see this pattern repeated a lot.
So just an idea, maybe for future things:

```rust
trait AddIoContextExt<T> {
    fn add_io_context(self, path: Path) -> Result<T, Error>;
}

impl<T> AddIoContext<T> for Result<T, std::io::Error> {
    fn add_io_context(self, path: Path) -> Result<T, Error> {
        self.map_err(|e| Error::IO(e, path))
    }
}


let len = hashes_file
    .metadata()
    .add_io_context(path.clone())?;

```

It does not reduce the lines of code but to me looks simplier

Reply:
Or maybe even better:

```rust
trait AddErrContextExt<T> {
    type Context;

    fn add_err_context(self, context: Context) -> Result<T, Error>;
}

impl<T> AddErrContext<T> for Result<T, std::io::Error> {
    type Context = Path;

    fn add_err_context(self, path: Self::Context) -> Result<T, Error> {
        self.map_err(|e| Error::IO(e, path))
    }
}
```

So that it becomed usefull not only for `Error::IO`

Reply:
I probaly just took the idea from [anyhow's `with_context()`](https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.with_context)